src/v/security/oidc_service.cc

```diff
@@ -0,0 +1,226 @@
+
+struct service::impl {
+    impl(
+      config::binding<std::vector<ss::sstring>> sasl_mechanisms,
+      config::binding<std::vector<ss::sstring>> http_authentication,
+      config::binding<ss::sstring> discovery_url,
+      config::binding<ss::sstring> token_audience,
+      config::binding<std::chrono::seconds> clock_skew_tolerance)
+      : _verifier{}
+      , _sasl_mechanisms{std::move(sasl_mechanisms)}
+      , _http_authentication{std::move(http_authentication)}
+      , _discovery_url{std::move(discovery_url)}
+      , _token_audience{std::move(token_audience)}
+      , _clock_skew_tolerance{std::move(clock_skew_tolerance)} {
+        _sasl_mechanisms.watch([this]() { return update(); });
+        _http_authentication.watch([this]() { return update(); });
+        _discovery_url.watch([this]() { return update(); });
+    }
+
+    ss::future<> start() { co_await update(); }
+    ss::future<> stop() { return _gate.close(); }
+
+    ss::future<> update() {
+        auto enabled = absl::c_any_of(
+                         _sasl_mechanisms(),
+                         [](auto const& m) { return m == "OAUTHBEARER"; })
+                       || absl::c_any_of(
+                         _http_authentication(),
+                         [](auto const& m) { return m == "OIDC"; });
+        if (!enabled) {
+            co_return;
+        }
+
+        auto holder = _gate.hold();
+
+        co_await update_metadata();
+        if (!_issuer.has_value()) {
+            vlog(seclog.error, "Failed to retrieve issuer from metadata");
+            co_return;
+        }
+        if (!_jwks_url.has_value()) {
+            vlog(seclog.error, "Failed to retrieve jwks_url from metadata");
+            co_return;
+        }
+
+        co_await update_jwks();
+        if (!_jwks) {
+            vlog(seclog.error, "Error updating keys: Keys not found");
+            co_return;
+        }
+
+        auto res = _verifier.update_keys(*_jwks);
+        if (res.has_error()) {
+            vlog(seclog.error, "Error updating keys: {}", res.assume_error());
+            co_return;
+        }
+    }
+
+    ss::future<> update_metadata() {
+        ss::sstring response_body = co_await make_request(_discovery_url());
+        auto metadata = oidc::metadata::make(response_body);
+        if (metadata.has_error()) {
+            vlog(
+              seclog.warn,
+              "invalid response from discovery_url: {}, errc: {}",
+              response_body,
+              metadata.error());
+            co_return; // errc::invalid_credentials;
+        }
+
+        vlog(seclog.trace, "metadata response_body: {}", response_body);
+        _issuer.emplace(metadata.assume_value().issuer());
+        _jwks_url.emplace(metadata.assume_value().jwks_uri());
+    }
+
+    ss::future<> update_jwks() {
+        if (!_jwks_url.has_value()) {
+            vlog(seclog.warn, "jwks_uri is not set");
+        }
+
+        auto response_body = co_await make_request(*_jwks_url);
```

Comment:
> Did you want to return an error if `_jwks_url` is empty?

There are many ways it can fail, but I wouldn't expect the response from the OIDC provider to have an empty `jwks_url`.

Reply:
Right it's just that you directly use it on line 167 below regardless of it having a value or not

Reply:
A return was missing, added.